src/plugins/intel_npu/src/plugin/include/metadata.hpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace intel_npu {
+
+/**
+ * @brief Magic bytes used for identifying NPU blobs.
+ */
+constexpr std::string_view MAGIC_BYTES = "OVNPU";
+
+/**
+ * @brief Returns a uint32_t value which represents two uint16_t values concatenated.
+ *
+ * @param major Major version.
+ * @param minor Minor version.
+ * @return Major and minor versions concatenated into one single uint32_t value.
+ */
+constexpr uint32_t make_version(uint16_t major, uint16_t minor) {
+    return major << 16 | (minor & 0x0000ffff);
+}
+
+/**
+ * @brief Gets the major version.
+ *
+ * @param version Metadata version.
+ * @return Major version.
+ */
+constexpr uint16_t get_major(uint32_t version) {
+    return static_cast<uint16_t>(version >> 16);
+}
+
+/**
+ * @brief Gets the minor version.
+ *
+ * @param version Metadata version.
+ * @return Minor version.
+ */
+constexpr uint16_t get_minor(uint32_t version) {
+    return static_cast<uint16_t>(version);
+}
+
+/**
+ * @brief List of supported version formats.
+ */
+constexpr uint32_t METADATA_VERSION_1_0{make_version(1, 0)};
+
+/**
+ * @brief Current metadata version.
+ */
+constexpr uint32_t CURRENT_METADATA_VERSION{METADATA_VERSION_1_0};
+
+struct OpenvinoVersion {
+private:
+    std::string version;
+    uint32_t size;
+
+public:
+    OpenvinoVersion(const std::string& version);
+
+    /**
+     * @brief Reads version data from a stream.
+     *
+     * @param stream Input stream to read from.
+     */
+    void read(std::istream& stream);
+
+    /**
+     * @brief Gets the version string.
+     *
+     * @return Version as std::string_view.
+     */
+    std::string_view get_version();
+
+    /**
+     * @brief Gets the size of version string as reference.
+     * Needed for Metadata.write() method.
+     * @return Reference to the size.
+     */
+    uint32_t& get_size();
+};
+
+struct MetadataBase {
+    /**
+     * @brief Reads metadata from a stream.
+     *
+     * @param stream Input stream to read from.
+     */
+    virtual void read(std::istream& stream) = 0;
+
+    /**
+     * @brief Writes metadata to a stream.
+     *
+     * @param stream Output stream to write to.
+     */
+    virtual void write(std::ostream& stream) = 0;
+
+    /**
+     * @brief Checks if metadata is supported.
```

Comment:
I think this nice comment should go to struct Metadata<METADATA_VERSION_1_0>::is_compatible(), right?
The exact purpose is for each implementation to implement its own checks. METADATA_VERSION_1_1 might check something else.